pad_server/src/controller.c

```diff
@@ -100,8 +103,34 @@ static void controller_cleanup(void *arg) { controller_disconnect((controller_t
  * @param n The number of bytes to receive. Must be less than or equal to the length of `buf`.
  * @return The number of bytes read. 0 indicates no more bytes, -1 indicates an error and `errno` will be set.
  */
-static ssize_t controller_recv(controller_t *controller, void *buf, size_t n) {
-    return recv(controller->client, buf, n, 0);
+static ssize_t controller_receive(controller_t *controller, void *buf, size_t n) {
+    ssize_t bread = 0;
+    size_t total_read = 0;
+
+    while (total_read < n) {
```

Comment:
Is the reason for this loop because you were observing receive calls where only partial packets were received?

Reply:
No, I tried using iovecs to send from the client and iovecs to receive from the server, at the time I wasn't aware about the tricks of a stream protocol like tcp.
So I reverted back the pad_server receiving logic to what there is now, which is that while loop